### ISSUE TITLE
refactor switch statement in command.hpp

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/stan#392]
I wanted to call out a design issue that's come up in the
command-line interface.  The code in question is this
from gm/command.hpp:

```
  ...

        } else if (algo->value() == "hmc") {
          
          int engine_index = 0;
          list_argument* engine = dynamic_cast<list_argument*>(algo->arg("hmc")->arg("engine"));
          if (engine->value() == "static") {
            engine_index = 0;
          } else if (engine->value() == "nuts") {
            engine_index = 1;
          }
          
          int metric_index = 0;
          list_argument* metric = dynamic_cast<list_argument*>(algo->arg("hmc")->arg("metric"));
          if (metric->value() == "unit_e") {
            metric_index = 0;
          } else if (metric->value() == "diag_e") {
            metric_index = 1;
          } else if (metric->value() == "dense_e") {
            metric_index = 2;
          }

          int sampler_select = engine_index + 10 * metric_index + 100 * static_cast<int>(adapt_engaged);
          switch (sampler_select) {
            case 0: {
              typedef stan::mcmc::unit_e_static_hmc<Model, rng_t> sampler;
              sampler_ptr = new sampler(model, base_rng, &std::cout, &std::cout);
              if (!init_static_hmc<sampler>(sampler_ptr, algo)) return 0;
              break;
            }
            case 1: {        
              typedef stan::mcmc::unit_e_nuts<Model, rng_t> sampler;
              sampler_ptr = new sampler(model, base_rng, &std::cout, &std::cout);
              if (!init_nuts<sampler>(sampler_ptr, algo)) return 0;
              break;
            }
            ...
```

I have a very strong preference for more direct tests, that is,
not going through construction of an integer and then a switch.
This would replace all the above code with:

```
  } else if (algo->value() == "hmc") {
    list_argument* metric = dynamic_cast<list_argument*>(algo->arg("hmc")->arg("metric"));
    std::string metric_config = metric->value();

    list_argument* engine = dynamic_cast<list_argument*>(algo->arg("hmc")->arg("engine"));
    std::string engine_config = engine->value();

    if (metric_config == "unit_e" && engine_config == "static" && !adapt_engaged) {
      typedef stan::mcmc::unit_e_static_hmc<Model, rng_t> sampler;
      sampler_ptr = new sampler(model, base_rng, &std::cout, &std::cout);
      if (!init_static_hmc<sampler>(sampler_ptr, algo)) return 0;
    } else if (metric_config == "unit_e" && engine_config == "nuts" && !adapt_engaged) {
      typedef stan::mcmc::unit_e_nuts<Model, rng_t> sampler;
      sampler_ptr = new sampler(model, base_rng, &std::cout, &std::cout);
      if (!init_nuts<sampler>(sampler_ptr, algo)) return 0;
    }
```

I have a very strong preference for option 2.   But an even
better version was suggested by Daniel as option 3:

```
    std::string get_engine(const list_argument* algo) {
      return algo->arg("hmc")->arg("engine")->value();
    }

    std::string get_metric(const algorithm& algo) {
      ... as above ...
    }


    std::string get_adapt_engaged(const list_argument* algo) {
       ... as above ...
    }

    bool is_unit_e_static(const algoright& algo) {
      return "unit_e" == get_metric(algo)
             && "static" == get_engine(algo)
             && "engaged" == get_adapt_engaged(algo);
    }
```

you need one of these for each base class you're going to construct (here, for each metric, engine, and adaptation strategy).  Then the switch becomes
even more readable:

```
    if (is_unit_e_static(algo)) {
      typedef stan::mcmc::unit_e_static_hmc<Model, rng_t> sampler;
      sampler_ptr = new sampler(model, base_rng, &std::cout, &std::cout);
      if (!init_static_hmc<sampler>(sampler_ptr, algo)) return 0;
    } else if (is_unit_e_nuts(algo)) {
      ...
```

The general rule is to favor named functions for everything.  Then *test each of these functions with a simple unit test*.  

The problem with option 1 above is that there's absolutely no way to
test the components of the embedded switch except from a top-level
command test.

Benefit number 2 is that it's much easier for the next person to read
the intent of the code and maintain it.

**Comment by syclik** [comment]
@betanalpha, in another issue, you mentioned that the switch was pulled out in https://github.com/stan-dev/stan/tree/feature/issue-1361-writer_callback

Can we break those changes out into a separate pull request?

**Comment by syclik** [comment]
@betanalpha, never mind. We'll fix this with #1361.

**Comment by syclik** [comment]
Closing this. Will take care of it with #1751.
